packages/world/src/ResourceSelector.sol

```diff
@@ -44,7 +49,26 @@ library ResourceSelector {
    * Convert a selector to a string for more readable logs
    */
   function toString(bytes32 resourceSelector) internal pure returns (string memory) {
-    return string(abi.encodePacked(getNamespace(resourceSelector), "/", getFile(resourceSelector)));
+    bytes16 namespace = getNamespace(resourceSelector);
+    bytes16 file = getFile(resourceSelector);
+    return
+      string(
+        abi.encodePacked(
+          namespace == ROOT_NAMESPACE ? ROOT_NAMESPACE_STRING : namespace,
+          "/",
+          file == ROOT_FILE ? ROOT_FILE_STRING : file
+        )
+      );
+  }
+
+  /**
+   * Convert a selector to a trimmed string (no trailing `null` ASCII characters)
+   */
+  function toTrimmedString(bytes16 selector) internal pure returns (string memory) {
+    uint256 length;
+    for (; length < 16; length++) if (Bytes.slice1(selector, length) == 0) break;
+    bytes memory packedSelector = abi.encodePacked(selector);
+    return string(Bytes.setLength(packedSelector, length));
```

Comment:
just thinking aloud:

since we'll probably want to do more validating other than just null chars (e.g. slashes, underscores, non-alphanumeric, etc.), I wonder if its worth moving this logic to the client, and a Solidity function to just validate correctness?

Reply:
the reason we need this function is to create a custom function selector (`<namespace>_<file>_<func>()`) in `registerFunctionSelector`: the interface requires namespace and file to be `bytes16`, and it's possible to pass any string shorter than 16 bytes as argument, but the remaining bytes will be 0. But to create the function selector we need to trim those 0s, because they influence the hash and therefore the function selector. As an example: assume `namespace` is `bytes16("mud")`, `file` is `bytes16("increment")`, then `bytes4(keccak256(abi.encodePacked(namespace, "_", file, "_func()")))` isn't the same as `bytes4(keccak256(abi.encodePacked("mud_increment_func()")))`, (but `bytes4(keccak256(abi.encodePacked(trim(namespace), "_", trim(file), "_func()")))` is correct)

Reply:
a binary search for reference (would need to be modified for `bytes16`):

```solidity
function findNull(uint256 _str) internal pure returns (uint256 result) {
  result = 32;
  // safe because result = 32, and max sum of all subtractions = 32
  unchecked {
    // binary search for the \0 terminator
    if (_str & type(uint128).max == 0) {
      result -= 16;
      _str >>= 128;
    }
    if (_str & type(uint64).max == 0) {
      result -= 8;
      _str >>= 64;
    }
    if (_str & type(uint32).max == 0) {
      result -= 4;
      _str >>= 32;
    }
    if (_str & type(uint16).max == 0) {
      result -= 2;
      _str >>= 16;
    }
    if (_str & type(uint8).max == 0) {
      result -= 1;
      _str >>= 8;
    }
    // Another check is needed in case the string is not \0 terminated
    if (_str & type(uint8).max == 0) {
      result -= 1;
    }
  }
  return result;
}
```